File: ic_agi/audit_log.py
"""
IC-AGI — Append-Only Audit Log
================================

Every action in IC-AGI is logged in an append-only ledger.
Entries cannot be modified or deleted once written.

SECURITY RATIONALE:
- Immutability prevents post-hoc tampering of execution history.
- Every entry is chained (hash-linked) to its predecessor,
  creating a verifiable sequence.
- In production, this log would be replicated across multiple
  independent custodians and could use a Merkle tree for
  efficient verification.

MOCK NOTICE:
This implementation uses a simple in-memory list with SHA-256 chaining.
Production systems would use:
  - Distributed append-only stores (e.g., Trillian, immudb)
  - Hardware-backed tamper-evident logging
  - Cryptographic timestamping authorities
"""

import hashlib
import json
import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional


@dataclass
class AuditEntry:
    """
    A single entry in the audit log.
    
    Each entry contains:
    - A sequential index
    - A timestamp
    - The event data (who, what, when, where)
    - A hash of the previous entry (chain integrity)
    - A hash of this entry (for verification)
    """
    index: int
    timestamp: float
    data: Dict[str, Any]
    prev_hash: str
    entry_hash: str = ""

    def compute_hash(self) -> str:
        """
        Compute the SHA-256 hash of this entry.
        
        SECURITY RATIONALE:
        - Hash includes index, timestamp, data, and prev_hash.
        - Any modification to any field changes the hash.
        - Chain breaks are detectable by verifying hash linkage.
        """
        content = json.dumps({
            "index": self.index,
            "timestamp": self.timestamp,
            "data": self.data,
            "prev_hash": self.prev_hash
        }, sort_keys=True, default=str)
        return hashlib.sha256(content.encode()).hexdigest()


class AuditLog:
    """
    Append-only audit log with hash-chain integrity.
    
    SECURITY RATIONALE:
    - Entries can only be appended, never modified or deleted.
    - Each entry's hash depends on all previous entries.
    - Verification can detect any tampering.
    - The log can be replicated for redundancy.
    """

    def __init__(self):
        self._entries: List[AuditEntry] = []
        self._genesis_hash = hashlib.sha256(b"IC-AGI-GENESIS").hexdigest()

    def append_entry(self, data: Dict[str, Any]) -> AuditEntry:
        """
        Append a new entry to the audit log.
        
        SECURITY RATIONALE:
        - Timestamp is recorded at write time (not caller-supplied).
        - Entry is hash-chained to the previous entry.
        - Once appended, the entry is immutable.
        """
        prev_hash = self._entries[-1].entry_hash if self._entries else self._genesis_hash
        
        entry = AuditEntry(
            index=len(self._entries),
            timestamp=time.time(),
            data=data,
            prev_hash=prev_hash
        )
        entry.entry_hash = entry.compute_hash()
        
        self._entries.append(entry)
        return entry

    def verify_integrity(self) -> bool:
        """
        Verify the entire chain's integrity.
        
        SECURITY RATIONALE:
        - Recomputes every hash and checks linkage.
        - If ANY entry was modified, the chain breaks.
        - Returns False if tampering is detected.
        """
        for i, entry in enumerate(self._entries):
            # Verify hash correctness
            expected_hash = entry.compute_hash()
            if entry.entry_hash != expected_hash:
                return False

            # Verify chain linkage
            if i == 0:
                if entry.prev_hash != self._genesis_hash:
                    return False
            else:
                if entry.prev_hash != self._entries[i - 1].entry_hash:
                    return False

        return True

    def get_entries(
        self,
        source: Optional[str] = None,
        event: Optional[str] = None,
        limit: Optional[int] = None
    ) -> List[AuditEntry]:
        """
        Query audit entries with optional filters.
        
        SECURITY RATIONALE:
        - Read-only access to the log.
        - Filters are applied in-memory (no mutation risk).
        """
        results = self._entries

        if source:
            results = [e for e in results if e.data.get("source") == source]
        if event:
            results = [e for e in results if e.data.get("event") == event]
        if limit:
            results = results[-limit:]

        return results

    def __len__(self):
        return len(self._entries)

    def dump(self) -> List[Dict[str, Any]]:
        """Export the full log as a list of dictionaries (for inspection)."""
        return [
            {
                "index": e.index,
                "timestamp": e.timestamp,
                "data": e.data,
                "prev_hash": e.prev_hash[:16] + "...",
                "entry_hash": e.entry_hash[:16] + "..."
            }
            for e in self._entries
        ]

File: typescript/src/audit-log.ts
/**
 * AuditLog — Hash-chained append-only audit ledger.
 *
 * TypeScript port of saezbaldo/ic-agi `ic_agi/audit_log.py`.
 * See: https://github.com/saezbaldo/ic-agi
 *
 * Each entry is cryptographically linked to the previous one via
 * SHA-256 chaining. Any tampering with historical entries will
 * break chain verification.
 */

import { createHash } from 'crypto';

// Genesis sentinel — hardcoded string matching the Python implementation
const GENESIS_HASH = createHash('sha256').update('IC-AGI-GENESIS').digest('hex');

/**
 * Produce a deterministic JSON string: sorted keys, no spaces.
 * Matches Python's json.dumps(sort_keys=True, separators=(',', ':'), default=str).
 */
function stableStringify(value: unknown): string {
  if (value === null || typeof value !== 'object') {
    return JSON.stringify(value) ?? 'null';
  }
  if (Array.isArray(value)) {
    return `[${value.map(stableStringify).join(',')}]`;
  }
  const obj = value as Record<string, unknown>;
  const keys = Object.keys(obj).sort();
  const pairs = keys.map(k => `${JSON.stringify(k)}:${stableStringify(obj[k])}`);
  return `{${pairs.join(',')}}`;
}

export interface AuditEntry {
  /** Zero-based position in the chain. */
  index: number;
  /** Unix epoch seconds — set by append(), never caller-supplied. */
  timestamp: number;
  /** Arbitrary event data. */
  data: Record<string, unknown>;
  /** SHA-256 of the previous entry (or GENESIS_HASH for the first entry). */
  prevHash: string;
  /** SHA-256 of this entry's canonical form. */
  entryHash: string;
}

/**
 * Hash payload uses snake_case keys (`prev_hash`) intentionally to match the Python IC-AGI
 * implementation. Do NOT rename to camelCase (`prevHash`) — doing so would change the SHA-256
 * digest and break cross-language audit log verification.
 */
function computeEntryHash(entry: Omit<AuditEntry, 'entryHash'>): string {
  const content = stableStringify({
    data: entry.data,
    index: entry.index,
    prev_hash: entry.prevHash,
    timestamp: entry.timestamp,
  });
  return createHash('sha256').update(content).digest('hex');
}

export class AuditLog {
  private entries: AuditEntry[] = [];

  /**
   * Append a new entry to the chain.
   * The timestamp is set here — callers cannot supply it.
   */
  append(data: Record<string, unknown>): AuditEntry {
    const prevHash =
      this.entries.length > 0
        ? this.entries[this.entries.length - 1].entryHash
        : GENESIS_HASH;

    const partial: Omit<AuditEntry, 'entryHash'> = {
      index: this.entries.length,
      timestamp: Date.now() / 1000,
      data,
      prevHash,
    };
    const entry: AuditEntry = { ...partial, entryHash: computeEntryHash(partial) };
    this.entries.push(entry);
    return entry;
  }

  /**
   * Verify the integrity of the entire chain.
   * Recomputes every entry hash and checks chain linkage.
   * Returns false immediately on the first mismatch.
   */
  verify(): boolean {
    for (let i = 0; i < this.entries.length; i++) {
      const entry = this.entries[i];
      if (entry.entryHash !== computeEntryHash(entry)) return false;
      const expectedPrev = i === 0 ? GENESIS_HASH : this.entries[i - 1].entryHash;
      if (entry.prevHash !== expectedPrev) return false;
    }
    return true;
  }

  /** Return entries, optionally filtered by source, event type, or count limit. */
  getEntries(filter?: { source?: string; event?: string; limit?: number }): AuditEntry[] {
    let results = this.entries;
    if (filter?.source) results = results.filter(e => e.data['source'] === filter.source);
    if (filter?.event) results = results.filter(e => e.data['event'] === filter.event);
    if (filter?.limit) results = results.slice(-filter.limit);
    return results;
  }

  getEntry(index: number): AuditEntry | undefined {
    return this.entries[index];
  }

  get length(): number {
    return this.entries.length;
  }
}
